{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Fastqc: Fastq quality assessment step\n",
    "Version: 0.11.5\n",
    "CMD:\n",
    "`fastqc --extract <file_path> --outdir <output_dir_path>`\n",
    "\n",
    "\n",
    "Xenome:Isolation and removal of mouse derived DNA resulting in a fastq of human derived reads\n",
    "Version: 1.1.0\n",
    "CMD:\n",
    "`xenome classify\n",
    "    --log-file <logfile_file_path> \\\n",
    "    --tmp-dir <tmp_dir_path> \\\n",
    "    --verbose \\\n",
    "    --pairs \\\n",
    "    --fastq-in <read1_fq_file_path> \\\n",
    "    --fastq-in <read2_fq_file_path> \\\n",
    "    --prefix <prefix> \\\n",
    "    --graft-name <human_reference_fasta> \\\n",
    "    --host-name <mouse_reference_fasta> \\\n",
    "    --output-filename-prefix <prefix>`\n",
    "Notes: The mouse host reference used in Repositive's exome workflow is Mus_musculus.GRCm38.dna.primary_assembly.fa and the human reference genome is Homo_sapiens.GRCh38.dna.primary_assembly.fa. Also note that the above command is run within a python wrapper that deals with the issue that Xenome often doesn't exit properly, by self-terminating if no changes have been made to the size of file after a pre-specified period of time.\n",
    "\n",
    "\n",
    "BWA: Alignment of read pairs to the reference genome, producing an alignment file SAM\n",
    "Version:0.7.17\n",
    "Samtools: Conversion of a SAM file to a BAM file\n",
    "Version:1.9\n",
    "CMD:\n",
    "```bwa mem -t <threads>\n",
    "    <reference_fasta>\n",
    "    <read1_fq_file_path>\n",
    "    <read2_fq_file_path>\n",
    "    2> <logfile_file_path>  |  samtools sort -O BAM -threads -o <output_bam_file_path>```\n",
    "Notes: The reference fasta used for alignment is Homo_sapiens.GRCh38.dna.primary_assembly.fa\n",
    "\n",
    "\n",
    "Picard MarkDuplicates: Identifies duplicate reads resulting from PCR duplication or Optical duplication during sequencing\n",
    "Version: 2.18.7\n",
    "CMD:\n",
    "`java -Djava.io.tmpdir=<tmp_dir_path> \\\n",
    "    -jar picard.jar MarkDuplicates \\\n",
    "    I=<input_bam_file_path> \\\n",
    "    O=<output_bam_file_path> \\\n",
    "    M=<metrics_file_path> \\\n",
    "    CREATE_INDEX=true > <log_file_path>`\n",
    " \n",
    "\n",
    "Picard AddOrReplaceReadGroups: Identifies duplicate reads resulting from PCR duplication or Optical duplication during sequencing\n",
    "Version: 2.18.7\n",
    "`java -jar picard.jar AddOrReplaceReadGroups \\\n",
    "    I=<input_bam_file_path> \\\n",
    "    O=<output_bam_file_path> \\\n",
    "    RGID=<read_group_platform_id> \\\n",
    "    RGPU=<read_group_platform_unit> \\\n",
    "    RGLB=<read_group_library> \\\n",
    "    RGPL=<read_group_platform e.g. illumina> \\\n",
    "    RGSM=<read_group_sample_name> \\\n",
    "    CREATE_INDEX=true > <log_file_path>`\n",
    "    \n",
    "GATK Base quality score recalibration: Detects systematic errors made by the sequencer when it estimates the quality score of each base call and marks these bases for recalibration\n",
    "Version: 4.0.5.1 \n",
    "CMD:\n",
    "`java -jar gatk-package-4.0.5.1-local.jar  BaseRecalibrator \\\n",
    "    --input <input_bam_file_path> \\\n",
    "    --known-sites GATK/hg38/1000G_omni2.5.hg38.ebi_chr.vcf.gz \\\n",
    "    --known-sites GATK/hg38/1000G_phase1.snps.high_confidence.hg38.ebi_chr.vcf.gz \\\n",
    "    --known-sites GATK/hg38/Axiom_Exome_Plus.genotypes.all_populations.poly.hg38.ebi_chr.vcf.gz \\\n",
    "    --known-sites GATK/hg38/dbsnp_146.hg38.ebi_chr.vcf.gz \\\n",
    "    --reference <reference_fasta> \\\n",
    "    --output <output_recalibration_report_path> > <log_file_path>`\n",
    "Notes: The reference fasta used for alignment is hs38DH.ebi_chr.fa\n",
    "\n",
    "\n",
    "GATK Apply BQSR: Applies the recalibration of base quality score for those bases marked in the previous step\n",
    "Version: 4.0.5.1\n",
    "CMD:\n",
    "`java -jar gatk-package-4.0.5.1-local.jar  ApplyBQSR \\\n",
    "    --create-output-bam-index true \\\n",
    "    --input <input_bam_file_path> \\\n",
    "    --bqsr-recal-file <input_recalibration_report_file_path> \\\n",
    "    --reference <reference_fasta> \\\n",
    "    --output <output_bam_file_path> > <log_file_path>`\n",
    "\n",
    "\n",
    "Samtools merge: Merges multiple bam files into a single bam, e.g. this is done when there are multiple lane runs for a sample.\n",
    "Version: 1.9\n",
    "CMD:\n",
    "`samtools merge -O BAM <output_bam_file_path> \\\n",
    "    <lane1_input_bam_file_path> <lane2_input_bam_file_path> <lane3__bam_file_path> <lane4_bam_file_path> \\\n",
    "    > <log_file_path>`\n",
    "\n",
    "\n",
    "Samtools index: Constructs an index file (a .bai file) for a bam file\n",
    "Version: 1.9\n",
    "CMD:\n",
    "`samtools index <input_bam_file_path>  <output_bai_file_path>`\n",
    "\n",
    "\n",
    "Samtools stats: Produces a stats file for each bam file, specifying the a number of read and base characteristics\n",
    "Version: 1.9\n",
    "CMD:\n",
    "`samtools stats <input_bam_file_path>  <output_stats_file_path>`\n",
    "\n",
    "\n",
    "GATK Haplotype caller: A variant caller, that identifies all the sites in the bam file that are variable from the reference genome\n",
    "Version: 4.0.5.1\n",
    "CMD:\n",
    "`java -jar gatk-package-4.0.5.1-local.jar  HaplotypeCaller \\\n",
    "    --reference <reference_fasta> \\\n",
    "    --input <input_bam_file_path> \\\n",
    "    --read-index <input_bai_file_path> \\\n",
    "    --TMP_DIR <tmp_dir_file_path> \\\n",
    "    --output <output_vcf_file_path> > <log_file_path>`\n",
    "Notes: The reference fasta used for alignment is hs38DH.ebi_chr.fa\n",
    "\n",
    "\n",
    "haplofilter: In house python script that filters vcf to remove calls with a quality score below the minQual\n",
    "Version: 1.0\n",
    "CMD:\n",
    "`haploFilter.py --inputVcf <input_vcf_file_path> \\\n",
    "    --minQual <quality_threshold e.g. 30> \\\n",
    "    --outputVcf <output_vcf_file_path>`\n",
    "Notes: In our exome pipeline the minimum quality threshold specified is 30.\n",
    "\n",
    "\n",
    "Bcftools stats: Produces a stats file for each vcf file charactertising the different variant types.\n",
    "Version: 1.9\n",
    "CMD:\n",
    "`bcftools stats <input_bam_file_path>  <output_stats_file_path>`   \n",
    "\n",
    "\n",
    "Variant Effect Predictor: Produces annotation for the variants present in the VCF.\n",
    "Version: 1.9\n",
    "CMD:\n",
    "`ensembl-vep-release-90/vep --offline \\\n",
    "    --hgvs \\\n",
    "    --fasta <reference_fasta> \\\n",
    "    --af_gnomad \\\n",
    "    -custom GRCh38/gnomad.genomes.r2.0.1.sites.GRCh38.noVEP.vcf.gz,\\\n",
    "    gnomADg,vcf,exact,0,AF_AFR,AF_AMR,AF_ASJ,AF_EAS,AF_FIN,AF_NFE,AF_OTH \\\n",
    "    -i <input_vcf_file_path> \\\n",
    "    -o <output_vep_annotations_file_path> \\\n",
    "    --dir_cache /<vep_cache_dir_path> > <log_file_path>`\n",
    "\n",
    "\n",
    "Merge VEP annotations: In house python script that adds the annotation to a VCF file, with each annotation value being placed in its own column.\n",
    "Version: 1.0\n",
    "CMD:\n",
    "`merge_vep.py --vep <input_vep_annotations_file_path> \\\n",
    "    --vcf <input_vcf_file_path> \\\n",
    "    --genes gene_translations.tsv \\\n",
    "    --mode local \\\n",
    "    --out <output_annotated_vcf_file_path>`\n",
    "Notes: The gene_translations.tsv is sourced from the Ensembl -> HGNC gene ID table"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
